Google Cloud Storage (GCS)

petedannemann opened the pull request:
#### Google Cloud Storage (GCS) Support

#### Motivation
- Adds GCS Support #198 

#### Checklist

Before you create the PR, please make sure you have:

- [x] Picked a concise, informative and complete title
- [x] Clearly explained the motivation behind the PR
- [x] Linked to any existing issues that your PR will be solving
- [x] Included tests for any new functionality
- [x] Checked that all unit tests pass

We will need to figure out how we plan to deal with integration testing on GCP. Would RaRe be willing to host the bucket? We will need to update Travis to include those tests if so.

EDIT: Removed comment about the testing timeout issue. Since fixing the memory issue with reads, it has gone away.

mpenkov commented on README.rst:
```diff
@@ -174,6 +175,14 @@ More examples
+
+    # stream content *into* GCS (write mode):
+    with open('gs://my_bucket/my_file.txt') as fout:
+        fout.write(b'hello world')
```
Why are you writing bytes here? Isn't `fout` open in text mode? Also, you're not specifying write mode here. This example looks broken.

mpenkov commented on integration-tests/test_gcs.py:
```diff
@@ -0,0 +1,120 @@
+
+
+def write_read(key, content, write_mode, read_mode, encoding=None, **kwargs):
+    with smart_open.smart_open(key, write_mode, encoding=encoding, **kwargs) as fout:
```
Please use the `open` function, here and everywhere. The `smart_open` function is deprecated.

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,524 @@
+
+logger = logging.getLogger(__name__)
+
+READ_BINARY = 'rb'
```
These can probably be internal.

Please review all module-level variables and functions. If they aren't needed from outside this module, then make them internal. If they are needed outside, add descriptive docstrings.

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,524 @@
+
+_RESUME_INCOMPLETE = 308
+
+BINARY_NEWLINE = b'\n'
```
If these are not internal, add docstrings for them.

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,524 @@
+START = 0
+CURRENT = 1
+END = 2
+WHENCE_CHOICES = [START, CURRENT, END]
```
This can be a tuple.

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,524 @@
+        The buffer size to use when performing I/O.
+    min_part_size: int, optional
+        The minimum part size for multipart uploads.  For writing only.
+    client: object, optional
```
Be more precise about expected data types:

```suggestion
    client: google.cloud.storage.Client, optional
```

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,524 @@
+    """Read an GCS blob."""
+    def __init__(self,
+                 gcs_blob,  # type: storage.Blob
+                 ):
```
This can fit on a single line, no need to break across multiple lines.

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,524 @@
+        raise NotImplementedError('GCS support for mode %r not implemented' % mode)
+
+
+class RawReader(object):
```
This looks like a useless class to me. You can replace it easily with:

```python
alternative = io.BytesIO(gcs_blob.download_as_string())
```

and it will behave the same way.

BTW, how large are these blobs? Is it acceptable to store them entirely in memory? I suspect the answer is no.

petedannemann replied:
Great suggestion, RawReader has been removed

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,524 @@
+        self._blob = bucket.get_blob(key)   # type: storage.Blob
+        self._size = self._blob.size if self._blob.size else 0
+
+        self._raw_reader = RawReader(self._blob)
```
This loads the whole thing into memory, is this really what we want?

mpenkov commented on smart_open/smart_open_lib.py:
```diff
@@ -831,6 +846,16 @@ def _unquote(text):
+
+    bucket_id, blob_id = parsed_uri.netloc, parsed_uri.path[1:]
+
+    return Uri(
```
This can fit on one line.

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,524 @@
+                self._eof = True
+
+
+class SeekableBufferedInputBase(BufferedInputBase):
```
Why do you need two separate classes?

1. BufferedInputBase
2. SeekableBufferedInputBase

Can't you get away with just one of them?

petedannemann replied:
Another great point, BufferedInputBase has been removed

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,524 @@
+import logging
+import sys
+
+from google.cloud import storage, exceptions
```
Please do this instead:

```suggestion
import google.cloud.exceptions
import google.cloud.storage
```

It makes your code more difficult to write, because it is more verbose.
However, it is easier to read, because the reader doesn't have to remember where `storage` and `exceptions` are coming from.
Readability is more important than writeability in this case.

petedannemann commented:
> Hi, thank you for your contribution.
> 
> I skimmed the reader part and left you some comments here and there. Please have a look.
> 
> My main concern is that you're loading the whole blob into memory when reading. That's what the `download_as_string` method appears to do. We definitely want to avoid this, because the remote object could be much larger than what can fit into memory.

Thanks for the review. A lot of these issues came from me reusing a lot of stuff based on how it was written in the s3 module, but it looks like that's not the approach we want to take. I'll work on addressing these issues, especially the one regarding memory

mpenkov commented:
That module evolved a bit over time, so you don't necessarily have to mimic everything it does.

Try to work around the memory issue (from memory, the S3 module doesn't suffer from this problem, so that particular part may provide some hints for you) and let me know when you're ready for another round of review.

Thanks!

piskvorky commented on integration-tests/test_gcs.py:
```diff
@@ -0,0 +1,120 @@
+import smart_open
+
+_GCS_BUCKET = os.environ.get('SO_GCS_BUCKET')
+_GCS_URL = 'gs://' + _GCS_BUCKET
```
Why `gs://` when the service is GCS?

petedannemann replied:
`gs://` is the most common scheme for Google Cloud Storage, but `gcs://` is also used occasionally

piskvorky replied:
Is there some specification for this? Where does this convention come from, where is it used?

petedannemann replied:
[Google's tools](https://cloud.google.com/bigquery/docs/loading-data-cloud-storage#gcs-uri) (gsutil, Google Cloud Storage console uris) all use `gs`, I actually had not seen `gcs` before the issue that requested this feature, and think maybe we should drop `gcs` for simplicity.

piskvorky replied:
OK, Google itself = the originator of the protocol is a source that's authoritative enough :)

petedannemann replied:
I agree, it's very confusing because they use the GCS acronym to refer to the Google Cloud Storage service. I wouldn't be surprised is their reasoning was one less letter to send in network calls, sounds about right for Google haha.

To clarify, where did we land with this? Just support `gs` or both `gs` and `gcs`?

piskvorky replied:
I'm +1 on `gs` only, if that's the established standard. @mpenkov thoughts?

mpenkov replied:
Yes, `gs` is fine initially. If people complain, we can revisit later.

petedannemann commented:
@mpenkov I believe I have addressed the issues you brought up. Please review when you get the chance

mpenkov commented:
OK, I will review your changes during the week. In the meanwhile, the tests are failing for some reason in CI, please make them pass.

piskvorky commented on smart_open/gcs.py:
```diff
@@ -141,28 +125,34 @@ def seek(self, position):
-            binary = self._body[:size]
-        self._position += len(binary)
+            start, end = position, position + size
+            binary = self._blob.download_as_string(start=start, end=end)
```
What happens when trying to read past the end of the file? `end > self._size`

petedannemann replied:
`google.cloud.storage.Blob.download_as_string` appears to handle this for us. I added a test to verify.

```python
@maybe_mock_gcs
    def test_read_past_end(self):
        content = b'englishman\nin\nnew\nyork\n'
        put_to_bucket(contents=content)

        with smart_open.gcs.SeekableBufferedInputBase(BUCKET_NAME, BLOB_NAME) as fin:
            data = fin.read(100)

        self.assertEqual(data, content)
```

piskvorky commented on smart_open/smart_open_lib.py:
```diff
@@ -684,6 +693,8 @@ def _parse_uri(uri_as_string):
 
       * file
+      * gcs
+      * gs
```
One line (if the same protocol).

piskvorky commented on smart_open/tests/test_gcs.py:
```diff
@@ -0,0 +1,660 @@
+
+
+def mock_gcs(func):
+    '''Mocks the function and provides additional required arguments.'''
```
```suggestion
    '''Mock the function and provide additional required arguments.'''
```

PEP257: Docstrings in imperative mode.

piskvorky commented on smart_open/tests/test_gcs.py:
```diff
@@ -0,0 +1,660 @@
+
+@maybe_mock_gcs
+def setUpModule():
+    '''Called once by unittest when initializing this module.  Sets up the
```
```suggestion
    '''Called once by unittest when initializing this module.  Set up the
```

piskvorky commented on smart_open/tests/test_gcs.py:
```diff
@@ -0,0 +1,660 @@
+
+@maybe_mock_gcs
+def tearDownModule():
+    '''Called once by unittest when tearing down this module.  Empties and
```
```suggestion
    '''Called once by unittest when tearing down this module. Empty and
```

piskvorky commented on setup.py:
```diff
@@ -59,14 +59,15 @@ def read(fname):
     'boto >= 2.32',
     'requests',
     'boto3',
+    'google-cloud-storage',
```
How heavy is this dependency? (download size, subdependencies, any compilation?, etc)

petedannemann replied:
The wheel is [72.7 kB on pypi](https://pypi.org/project/google-cloud-storage/#files). It has [three other GCP dependencies](https://github.com/googleapis/google-cloud-python/blob/master/storage/setup.py), I do not see any compilation required. I think optional installation for smart_open is a good idea, and can move it into `extras_require` in the setup.py

piskvorky replied:
Nah. If it's light (and reasonably stable), I'd prefer a direct dependency, to keep things simple.

piskvorky commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,521 @@
+_READ_BINARY = 'rb'
+_WRITE_BINARY = 'wb'
+_MODES = (_READ_BINARY, _WRITE_BINARY)
+"""Allowed I/O modes for working with GCS."""
```
Why are these docstrings? I'd expect a comment *above* each variable, instead of a docstring *below* it.

mpenkov replied:
Sphinx picks them up that way.

Not that we use Sphinx for this project, but it's a convention.

petedannemann replied:
This was done in response to @mpenkov's [comment](https://github.com/RaRe-Technologies/smart_open/pull/404#discussion_r363068746). The convention used in the S3 module is docstrings below the constant.

piskvorky replied:
OK, but I find that thoroughly confusing. In the following block of code, which comment relates to which variable?

```python

START = 0
"""Seek to the absolute start of a GCS file"""
CURRENT = 1
"""Seek relative to the current positive of a GCS file"""
END = 2
"""Seek relative to the end of a GCS file"""
_WHENCE_CHOICES = (START, CURRENT, END)

```

petedannemann commented:
> OK, I will review your changes during the week. In the meanwhile, the tests are failing for some reason in CI, please make them pass.

I made some fixes and now the only error is from Python 2.7 not recognizing the `test_gcs.py` file when it runs `python setup.py test`. I'm pretty stumped on why this is happening, any ideas?

petedannemann commented:
> OK, I will review your changes during the week. In the meanwhile, the tests are failing for some reason in CI, please make them pass.

The tests are passing, but the build for Python 2.7 is failing when it runs `python setup.py test` due to the error below:
```bash
>> AttributeError: 'module' object has no attribute 'test_gcs'
```

Either of you know how to get Python 2.7 to discover my new test file?

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,523 @@
+        elif size < 0:
+            from_buf = self._read_from_buffer()
+            self._current_pos = self._size
+            return from_buf + self._raw_reader.read()
```
You don't need to declare variables here.

```suggestion
            return self._read_from_buffer() + self._raw_reader.read()
```

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,523 @@
+
+    def terminate(self):
+        """Do nothing."""
+        pass
```
Is this OK?

Theoretically, you can continue reading after calling terminate, which doesn't sound very useful. Why is this function even needed?

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,523 @@
+                self._eof = True
+
+    def __str__(self):
+        return "(%s, %r, %r)" % \
```
One line.

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,523 @@
+        self._credentials = self._client._credentials
+        self._bucket = self._client.bucket(bucket)  # type: google.cloud.storage.Bucket
+        self._blob = self._bucket.blob(blob)  # type: google.cloud.storage.Blob
+        assert min_part_size % _MIN_MIN_PART_SIZE == 0, 'min part size must be a multiple of 256KB'
```
You really mean this, right?

```suggestion
        assert min_part_size % _REQUIRED_CHUNK_MULTIPLE == 0, 'min part size must be a multiple of 256KB'
```

mpenkov replied:
Also, you need to check `min_part_size >= _MIN_MIN_PART_SIZE`.

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,523 @@
+        logger.debug("closing")
+        if self._total_size == 0:  # empty files
+            self._upload_empty_part()
+        if self._buf.tell():
```
Do you need to clean up other class members here?

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,523 @@
+
+        self._total_size = 0
+        self._total_parts = 0
+        self._buf = io.BytesIO()
```
I think a better name would be:

```suggestion
        self._current_part = io.BytesIO()
```

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,523 @@
+        content_length = self._buf.tell()
+        start = self._total_size - content_length
+        stop = self._total_size - 1
+        if stop - start + 1 == content_length:
```
What is this fidgeting here?

LHS:

- `stop - start + 1`
- `((self._total_size - 1) - (self._total_size - content_length) + 1` (substituting)
- `self._total_size - 1 - self._total_size + content_length + 1` (expand parentheses)
- `content_length` (simplify, cancel out)

So, the LHS is always equal to `content_length`, which is equal to RHS.

What am I missing here?

mpenkov commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,523 @@
+            'Content-Range': _make_range_string(start, stop, end)
+        }
+        data = self._buf
+        response = self._session.put(self._resumeable_upload_url, data=data, headers=headers)
```
You don't need to declare `data` just to use it in the line below.

```suggestion
        response = self._session.put(self._resumeable_upload_url, data=self._buf, headers=headers)
```

mpenkov commented on smart_open/tests/test_gcs.py:
```diff
@@ -0,0 +1,659 @@
+            self._create_if_not_exists()
+
+    def create_resumable_upload_session(self):
+        resumeable_upload_url = self.RESUMABLE_SESSION_URI_TEMPLATE.format(
```
Please don't use .format, we avoid it in our source code. Use `formatstr % dict(foo='bar')`.

mpenkov commented on smart_open/tests/test_gcs.py:
```diff
@@ -0,0 +1,659 @@
+
+
+class FakeBlob(object):
+    RESUMABLE_SESSION_URI_TEMPLATE = (
```
Move this to module scope.

mpenkov commented on smart_open/tests/test_gcs.py:
```diff
@@ -0,0 +1,659 @@
+    warnings.filterwarnings("ignore", category=ResourceWarning, message="unclosed.*<ssl.SSLSocket.*>")  # noqa
+
+
+class FakeBucket(object):
```
It looks like you're implementing your own mocks. Is there really no library to mock GCS? (looks like [there isn't](https://github.com/googleapis/google-cloud-python/issues/5816)).

If so, then we need to write tests for these mocks as well, to ensure they're behaving as expected. The risk is that the mock implementation is buggy in a particular way that hides bugs in our GCS tests.

petedannemann replied:
I was unable to find a mocking library for GCS. I will write tests for the mocks. Do you think these should all reside in the same file or should there be something like one file for the mocks, `gcs_utils.py` and another for the mock tests `test_gcs_utils`?

mpenkov replied:
Same file is fine.

petedannemann commented on integration-tests/test_gcs.py:
```diff
@@ -0,0 +1,120 @@
+    with smart_open.open(key, 'wb') as fout:
+        fout.write(one_megabyte_of_msgs)
+
+    actual = benchmark(read_length_prefixed_messages, key, 'rb', buffering=ONE_MIB)
```
> Shouldn't buffering be wrapped in a transport_kwargs dict? It's a GCS-only parameter, right?

Buffering is a kwarg for the built-in `open` function ([source](https://python-reference.readthedocs.io/en/latest/docs/functions/open.html)) and I thought it would be simpler for it to be named the same instead of making it specific to GCS. If you feel otherwise I can wrap it in transport_kwargs.

petedannemann replied:
> Did you try running these tests against a real GCS bucket?

Yes, I have ran both the integration and unit tests against a real bucket and they are currently passing.

petedannemann commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,523 @@
+_SUCCESSFUL_STATUS_CODES = (200, 201)
+
+
+def clamp(value, minval, maxval):
```
Changed so clamp is imported from S3 instead of copied here

petedannemann commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,523 @@
+
+        return len(b)
+
+    def terminate(self):
```
This was something I copied from the s3 module. I think having some way to abort an upload seems valuable, although it would probably be used very infrequently. I will document it

petedannemann commented:
@mpenkov @piskvorky I've been trying to think through the GCS mocks and imitate how moto works and unfortunately I think this will be a significantly larger endeavor than I originally thought. I also am worried about the maintainability of them in the long run and coupling them with the smart_open library doesn't seem like the right way of doing it. I see two options:

1. Develop a Python package for mocking GCS that lives outside of smart_open (this will probably take me months and will be a risk for smart_open)
2. Refactor all gcs tests to simply assert calls to the `google.cloud.storage library`. These tests will not be as good as the current ones, but will still yield some testing benefits

What do you think?

mpenkov commented:
I agree that 1) is probably too much for any of us to take on right now.

I think 2) is more of a step in the right direction. If the goal is to prevent regression, then your existing tests and mocks are sufficient. To prove correctness, we will have to use actual GCS buckets (integration testing).

petedannemann commented:
@mpenkov I retract my last comment and think the mocks are actually the best testing methodology for this situation. I think I got caught up in scope creep from looking at all of the functionality of moto, but the mocks here only need functionality as it relates to the tests they're being used with. I added tests for the mocks and am ready for another review.

mpenkov commented:
OK, I'll have a look within a week and let you know.

petedannemann commented on smart_open/gcs.py:
```diff
@@ -0,0 +1,530 @@
+        bucket_id,
+        blob_id,
+        mode,
+        buffering=DEFAULT_BUFFER_SIZE,
```
Changed to `buffer_size` to avoid this issue and mimic the s3 module

petedannemann commented:
> Looking very good. Left you some minor comments.
> 
> Also, you're missing tests for the top-level module (smart_open_lib). These don't have to be super-detailed, but you want to make sure that `smart_open.open("gs://bucket/key")` actually works.

Thanks @mpenkov! I added your comments and added those tests.

petedannemann commented:
Thanks for your help and patience!

wookayin commented:
How does it handle authentication? gcloud and gsutil are well-configured and working well but `open('gs://...')` does not work.

mpenkov commented:
@wookayin Can you please open a new ticket to describe your problem? Be sure to fill out the full issue template.

petedannemann commented:
@wookayin it uses Google's [google-cloud-storage](https://pypi.org/project/google-cloud-storage/) package so please refer to that package's documentation on how to configure authentication. The preferred method is to create a service account key file and set your GOOGLE_APPLICATION_CREDENTIALS environment variable to the path of that file.

mpenkov commented:
@petedannemann Is that documented anywhere, like the readme or howto.md? If not, it'd be good to add some documentation to point people in the right direction (the above comment is already a solid start).

petedannemann commented:
It is not. I will work on a PR to add this missing documentation soon and consider adding some examples of how you can use other authentication methods by passing in your own `google.cloud.storage.Client` object to transport_kwargs.
